facades/entrypoints.md

```diff
@@ -0,0 +1,239 @@
+
+**(c)** The facade should fetch network parameters (network config) needed for construction transactions (e.g. chain ID) in a lazy manner. For example, it may fetch them when the first transaction is created. This works fine for JavaScript, as well, since transaction-creation functions are async anyway.
+
+```
+class NetworkEntrypoint:
+    // The constructor is not captured by the specs; it's up to the implementing library to define it.
+    // For example, it can be parametrized with: a network provider URL and kind (proxy, API)
+    constructor(url: str, kind: Optional[str]);
+
+    verify_transaction_signature(transaction: Transaction): bool;
+    verify_message_signature(message: Message): bool;
+
+    // Fetches the account nonce from the network.
+    recall_account_nonce(address: IAddress): uint64;
+
+    // Function of the network provider, promoted to the facade.
+    send_transaction(transaction: Transaction): string;
+
+    // Function of the network provider, promoted to the facade.
+    send_transactions(transaction: Transaction); Tuple[int, List[string]];
+
+    // Generic function to await a transaction on the network.
+    await_completed_transaction(transaction_hash: string): TransactionOnNetwork;
+
+    // Access to the underlying network provider.
+    get_network_provider(): INetworkProvider;
+
+    // Access to the individual controllers.
+    create_smart_contract_controller(abi: Optional[Abi]): SmartContractController;
+    get_transfers_controller(): TransfersController;
+    get_token_management_controller(): TokenManagementController;
+    get_delegation_controller(): DelegationController;
+    get_relayed_controller(): RelayedController;
+    get_account_management__controller(): AccountManagementController;
+```
+
+## Examples
+
+### Transfer native amounts and / or custom tokens
+
+```
+entrypoint = MainnetEntrypoint();
+controller = entrypoint.get_transfers_controller()
+
+sender = Account.new_from_pem("alice.pem");
+sender.nonce = entrypoint.recall_account_nonce(sender.address);
+
+transaction = controller.create_transaction_for_transfer({
+    sender: sender,
+    nonce: sender.get_nonce_then_increment(),
+    receiver: Address.from_bech32("erd1..."),
+    native_amount: 1000000000000000000,
+    token_transfers: [];
+});
+
+// Broadcast transaction (option 1):
+transaction_hash = entrypoint.send_transaction(transaction);
+// Broadcast transaction (option 2):
+transaction_hash = entrypoint.get_network_provider().send_transaction(transaction);
+
+// Await transaction completion (option 1, await and specialized outcome parsing):
+parsed_outcome = controller.await_completed_transfer(transaction_hash);
+// Await transaction completion (option 2, simple await):
+transaction_on_network = entrypoint.await_completed_transaction(transaction_hash);
+```
+
+### Deploy a contract
+
+```
+entrypoint = MainnetEntrypoint();
+abi = Abi.load("adder.abi.json");
+controller = entrypoint.create_smart_contract_controller(abi);
```

Comment:
I believe this is what was bugging me:
`create_smart_contract_controller`
`get_transfers_controller`

They have the same purpose in the end but it's not **directly intuitive** which should be which and we got to this situation in which we actually ended up with both: one with get, another with create. 
Even though I am aware that this may happened by mistake because of it being a controller/factory/whatever in the lower layer, this kinda proves my point that I may look for any of those because of this pattern.
And makes me wish that I just could've just done `entrypoint.smart_contract_controller.get_me_a_call()`

Reply:
Other controllers - e.g. token management controller, delegation controller - do not have the dependency on "ABI" when created. Their lifetime is managed by the `entrypoint` itself. The entrypoint creates them when it's initialized, then provides their references when needed.

However, the controller for a smart contract is explicitly instantiated - thus the "create" verb. The caller (client code) provides the ABI.

Reply:
I see now, but only because you explained it. 
In honesty, this would annoy me quite badly because of never knowing what do I have to look for intuitively.

The mental effort for this is too much in my opinion if we're forcing the devs to differentiate a generic thing (creation of a tx) based on the way we implemented the controller. Sorry, but I don't see this as good design.

Reply:
Noted :+1: 

We can have this with _get_ instead of _create_, as well, and require the ABI when calling `create_transaction_for_execute`, `query` etc. - as it was the idea at first.

Reply:
Yeah... not very nice, I guess... I would rather instantiate the other controllers only when called, similarly to the contract controller, then call all of them `create` as well... I don't know. I give up.

Reply:
Adjusted to use `create_*` for controllers.